ENH: Added proper handling of nans for numpy.lib.function_base.median

empeeu opened the pull request:
to close issue #586.

Also added unit test.

This was previously PR #4287 , but I botched the rebase, so I just started over. 

Also related was PR #3908 and #4307

empeeu committed:
ENH: Added proper handling of nans for numpy.lib.function_base.median
to close issue #586.

Also added unit test.

empeeu committed:
MAINT: cleanup median nan check code
fix output argument, add warnings, fix style and add a note in the release notes.

shoyer commented:
This look like a useful patch to fix buggy behavior -- are there any remaining issues for merging this for numpy 1.10? @juliantaylor?

empeeu commented:
I rebased again since this was getting out of date. @jaimefrio can this be merged?

shoyer commented on numpy/lib/function_base.py:
```diff
@@ -2996,9 +2992,33 @@ def _median(a, axis=None, out=None, overwrite_input=False):
+                    rout = a.dtype.type(np.nan)
+        else:
+            for i in range(np.count_nonzero(n.ravel())):
+                warnings.warn("Invalid value encountered in median",
```
Issuing warnings in a loop seems like a bad idea, for two reasons:

1. It would just produce a lot of noise in logs (if anyone is not filtering all but the first warning)
2. Issuing warnings is actually very slow, so this could be a serious performance bottleneck. Consider:

```
In [10]: a = np.zeros((10000, 2))

In [11]: %timeit np.median(a, axis=0)
10000 loops, best of 3: 117 µs per loop

In [12]: %timeit for x in range(10000): warnings.warn("Invalid value encountered in median", RuntimeWarning)
/Users/shoyer/miniconda/envs/numpy-dev/bin/ipython:257: RuntimeWarning: Invalid value encountered in median
100 loops, best of 3: 12.3 ms per loop
```

homu commented:
:umbrella: The latest upstream changes (presumably #5977) made this pull request unmergeable. Please resolve the merge conflicts.

empeeu commented:
@shoyer I agree. Moved it outside the for loop, see #5753 

@homu I rebased again. Should now be able to merge automatically again on #5753. 

I'm closing this PR because #5753 handles both the median-nan and percentile-nan functions and I haven't had any response to my question April 7.
